### DISCUSSION TITLE
v4.2.0

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#182] [Announcements]
Feature-completeness and code quality is the focus of this release.

## What's Changed
**TL>DR**:
* Automated forecast updates that do not require an automation by @autoSteve and @BJReplay
* Granular dampening to dampen per half hour period and by site by @autoSteve and @isorin
* Dampening applied at forecast fetch and not to forecast history @autoSteve and @isorin
* Retrieve un-dampened forecast values using service call by @autoSteve (thanks @Nilogax)
* Get presently set dampening factors using service call by @autoSteve (thanks @Nilogax)
* Migration of un-dampened forecast to un-dampened cache on startup by @autoSteve
* Add site breakdown option for detailed forecasts by @autoSteve
* Add hard limit configuration to options by @autoSteve
* Suppress integration reload when many configuration options are changed by @autoSteve
* Translations of service call error responses by @autoSteve

**Automated forecast updates**

All users of this integration will have Home Assistant automation(s) set up to fetch forecast updates at desirable times of day. And that's great for the power users, who want an update just after midnight, then another strategic update pre-dawn, and then spread other updates up to API limit throughout the day (or not...).

If you don't need to do anything special, then head into the integration configure options and turn on 'Auto-update'. Disable or delete your automation. You can specify whether you want automated updates over twenty-four hours, or from sunrise to sunset. More in the readme.

**Granular dampening**

The only option to set dampening previously was at per-hour increments, and not the half hour forecast periods that Solcast provide to hobbyists. You can specify a total of 48 dampening values now, corresponding to one for every half-hourly forecast period, and do this per-site if needed. See the readme for more on this.

**Dampening at forecast fetch**

Previously, dampening was applied to all forecasts, both up-coming and all history, but those dampened values were not saved to retain historical differences in dampening that had changed over time.

The history of dampened forecasts is now preserved, and only newly fetched forecasts have dampening adjusted.

At 'day one' there will be no dampening applied to your history, but this will build over time.

**Retrieve un-dampened forecast values**

Because history is now dampened as-at the dampening values set at the time, automations might need un-dampened data to calculate future dampening based on consistent past differences in generation performance versus recent history. The 'get forecasts' service call is extended to allow this by specifying `undampened: true`.

At upgrade, the past fourteen days of undampened history will be saved, allowing this extended call to be used immediately. 

**Get currently set dampening values**

To avoid an automation needing to keep track of dampening set, the current dampening values can be retrieved with a new service call.

**Full spread of attributes breakdown by site**

Absolutely all sensor attributes may now be broken down by site, including the detailed half-hourly and hourly forecasts.

All attributes sent to Home Assistant may be configured in configure options. Turn them on or off to reduce clutter.

**Hard limit**

Hard limit is now in the configure options dialogue. Don't know what it does? Ask the readme. You probably don't need it.

**Suppression of integration re-load on options change**

Previously, should any configure option be changed then the integration would simply re-load. Now if the integration does not need to then it will not fully re-load.

**Full Changelog**: https://github.com/BJReplay/ha-solcast-solar/compare/v4.1.7...v4.2.0
**Changes since v4.1.9**: https://github.com/BJReplay/ha-solcast-solar/compare/v4.1.9...v4.2.0

**Comment by autoSteve** [comment]
Quite an exciting release, really. (And as @dtswk puts it hopefully boring, but I would hope of significant use and improvement.)

Notifying the usual early-adopting-and-reporting folk. @dtswk, @EddieSpaghettie, @ffm777, @gcoan, @hwb45, @iainfogg, @isorin, @jeffb1974, @ladaan, @Markyt77, @Nilogax, @Noodleyman, @ProphetOfDoom, @rholligan, @sh00t2kill, @Unlik and @wimdebruyn.

The upgrade went without a hitch for me.

Do ping if you would like to be off the list.

**Reply by ProphetOfDoom** [reply]
Updated to 4.2.0 with no dramas. Notting Hill rather than Alien.
You guys are doing an excellent job. I'm almost glad that Oziee decided to pass the baton.

**Comment by isorin** [comment]
Also running 4.2.0 from this morning. No issues so far.

**Comment by gcoan** [comment]
Upgraded from 4.1.8 to 4.2.0 (I didn't get a chance to go to 4.1.9 before it was superseded).

Messages in the log file about undampened history being migrated to a new json file, but otherwise no signs that there was anything new.  All seems to be working fine.

I'm interested in getting back to the conversation about auto-dampening based on historic data now that we have the more granular dampening capability, not sure where we are with that.

Since turning down the efficiency figure for my panels on Solcast a couple of months ago I am getting much better correlations of generation to forecast; it used to never be hitting the forecast but now its there or there abouts most days.  Interested to see if more analysis of historical data will enable a better fine tuning of the forecast via dampening factors

**Reply by BJReplay** [reply]
> I'm interested in getting back to the conversation about auto-dampening based on historic data now that we have the more granular dampening capability, not sure where we are with that.

@Nilogax has updated the discussion about dampening - his solution now uses the features from v4.1.9

**Reply by gcoan** [reply]
Thanks I saw that he'd posted that, I just couldn't see the code was updated in the github comments.
I'll go back and look at it properly

**Comment by EddieSpaghettie** [comment]
All good from my side as well!

**Comment by BJReplay** [comment]
![Screenshot_20241015_191618_Home Assistant.jpg](https://github.com/user-attachments/assets/9d056d29-0ac6-4b12-b502-cc8d7be76565)

Solid download rate over the last 24 hours.

This feels like a solid release.

Just don't go changing your API key 🤣.
